### ISSUE TITLE
[META] Private Addresses In Public

### ISSUE THREAD

**Issue by Stebalien** [ipfs/kubo#6932]
You may have noticed that IPFS currently announces _private_ addresses on the public DHT. While IPFS does support local discovery through MDNS, (a) MDNS isn't always perfectly reliable and (b) when trying to find a specific peer, we don't actually _query_ MDNS.

However, announcing private addresses to the public network is far from an ideal solution. There have been many discussions about this issue (#1226, #1173, #4343, #5418, #5511) and some deeper discussions in https://github.com/libp2p/go-libp2p/issues/436.

Problems include:

* Unnecessary network activity. Local dials are usually pretty cheap but they're still a waste.
* Triggers port scanning warnings (https://github.com/ipfs/go-ipfs/issues/4343). This is actually a rather complicated issue that includes _other_ issues like dialing too many nodes (should be fixed by incoming DHT improvements), accumulating too many observed addresses (should be fixed by signed peer routing records), and just the general "IPFS is a p2p application that needs to dial a lot of nodes".

However, we have _good news_!

1. Libp2p is introducing signed peer routing records. This means we're getting _atomic_ peer routing records that can't be spoofed, modified, extended, etc. This will land in 0.6.0.
2. Once that lands, libp2p will add a record "scope" (https://github.com/libp2p/go-libp2p/issues/793).
3. Go-ipfs 0.5.0 ~should (unless something changes) ship~ *ships* two DHTs. A public one and a private one: https://github.com/libp2p/go-libp2p/issues/803.

Once all three of those land, we should be able to restrict private addresses to the private DHT.

**Comment by RubenKelevra** [comment]
I hate to bump this, but it's quite a low hanging fruit and causing many headaches - as it requires manual interventions on every installed node to not push private IPs to the DHT and also not try to connect to private IPs received from the public DHT.

Is there a chance to get this on the roadmap for 0.9? :)

**Comment by drgorillamd** [comment]
Bump. Just got a warning from Hetzner...

**Comment by keesvanbochove** [comment]
> Bump. Just got a warning from Hetzner...

This is happening when you are running with server profile? Is it a similar issue like https://github.com/ipfs/go-ipfs/issues/5418#issuecomment-423743616?

**Comment by drgorillamd** [comment]
> 

I thought I did, but they removed my warning after config patching it, so now I guess the issue was on my side - sorry for the unnecessary bump

**Comment by fariszr** [comment]
> > > Bump. Just got a warning from Hetzner...
> > 
> > 
> > This is happening when you are running with server profile? Is it a similar issue like [#5418 (comment)](https://github.com/ipfs/go-ipfs/issues/5418#issuecomment-423743616)?
> 
> I thought I did, but they removed my warning after config patching it, so now I guess the issue was on my side - sorry for the unnecessary bump

Can you give more details on your setup ? Are you using a dedicated server ? Or cloud ?

I'm trying to install ipfs on a hetzner cloud server, I got a warning on my first try.

**Comment by Mouradif** [comment]
I just got my server terminated by hetzner due to this 😭

**Comment by acuteaura** [comment]
Hetzner's network team gets very unpleasant about these.

Maybe IPFS could check the first interface to see if that address is RFC1918 and then make local peer discovery opt-in at that point?

**Comment by acuteaura** [comment]
Update: I got ejected from Hetzner again; server profile does not fix this issue.

(Why? Because it doesn't ignore the CGNAT range and apparently people run IPFS on nodes with tailscale installed)

**Comment by ShadowJonathan** [comment]
@acuteaura IPFS does not rate limit connections at all, and this repo has not acknowledged deep-running issues like this for multiple years, you're SOL, I heavily recommend permanently dropping IPFS from your stack.

**Comment by sharp2448** [comment]
@ShadowJonathan do you know if there is any good alternative? 
Because I'm running graph node with subgraph, so I'm looking if I can use different approach which doesn't involve IPFS node ruining...

**Comment by sharp2448** [comment]
@Winterhuman Ok, sounds good. Thanks for info

**Comment by Jorropo** [comment]
@varuzam this is a different issue, please comment on https://github.com/ipfs/kubo/issues/8585

I am not able to reproduce the issue you are seeing, if you can help do so please comment on https://github.com/ipfs/kubo/issues/8585.

**Comment by ROBERT-MCDOWELL** [comment]
same here, my server at hetzner has been block because IPFS is send from port 4001 to many private ip range I never set in the config... what's the settings to stop this chaos?

**Comment by Jorropo** [comment]
@ROBERT-MCDOWELL this is a different issue, please comment on https://github.com/ipfs/kubo/issues/8585

**Comment by ROBERT-MCDOWELL** [comment]
duplicate issue.... https://github.com/ipfs/kubo/issues/8585

**Comment by TheRook** [comment]
@Stebalien Sending out RFC1918 addresses over the public network is always a bug - this should be forbidden and any client doing this needs to be patched.  Do you think this is a symptom of the client not knowing its external IP?

What if there was an RPC that just echos back the IP address revived on the socket handle?  With this RPC, the moment a peer has a connection they know how they routed out.  This could be used to identify a [double-NAT, which can still be traversed via UPnP](https://stackoverflow.com/questions/18865586/how-to-achieve-double-nat-port-forwarding-via-upnp-programmatically) as well as debug other issues like local peer discovery for peers who share a NAT but are on different network segments preventing mDNS, which is common in universities and corporate networks where multiple buildings  or floors share the same up-link.

Now if the client knows its IP address, they could hash this address (no port number) and see if anyone looks it up on the global DHT network.  If every client tries to lookup their own external IP - they'll find each-other and can exchange their RFC1918 local address so that they can connect directly over the local network which is much higher bandwidth.   Clients knowing about each other on the local network is also good as it prevents them from competing for resources, or clobbering each other's NAT routes.

**Comment by mahdiyari** [comment]
I think you guys need 3 more years to get this sorted out.

**Comment by Mouradif** [comment]
> I think you guys need 5 more years to get this sorted out.

Feel free to open a pull request

**Comment by TheRook** [comment]
Yeah, this feature already exists in I2p and cjdns.  Is IPFS going to adopt
open source solutions or are we going to wait 10 years for someone to
reinvent the wheel on this one?

**Comment by ShadowJonathan** [comment]
what "open source solutions"? Kubo is open source

**Comment by TheRook** [comment]
Freenet launched this feature 23 years ago...

Kubo isn't using open source.  With all due respect, Kubo is reinventing
wheels and then releasing the source code.

**Comment by ShadowJonathan** [comment]
Freenet isn't Kubo, sorry to break the news

**Comment by TheRook** [comment]
One can choose to stand on the shoulders of Giants.  It is best to choose
your battles, if this is already a solved problem - what ego is there to be
gained in slaving over a lesser solution?

Memento Mori. Choose your battles wisely.  I am posting here because
standing up for human rights is a choice that impacts the lives of others.

And when lives are in the line, shouldn't we use what is already working
well?

**Comment by TheRook** [comment]
@Jorropo 

There are a few open bugs that I believe are a symptom of the same problem, this issue included. 

libp2p's nat traversal fails when it is [behind multiple nats](https://github.com/libp2p/go-libp2p/issues/2074). Libp2p's hole punching is  unstable when [multiple clients attempt to use the same source port on the same subnet](https://github.com/libp2p/go-libp2p/issues/2059).

One problem is that the client cannot know what it's external IP address is by looking at it's local interface. Putting this on the human, means only smart humans will be able to use it - and that isn't a general solution.

What if we create two new RPCs that facilitate network connectivity.  Lets call them /net/ip and /net/ping.  A client can reach-out to a bootstrapped libp2p node and call /net/ip to get their external IP address so that external advertisements use an external ip address.

Now the next RPC, /net/ping is used, which will attempt a unique connect back to the same IP address where the call originated - but this RPC takes two arguments it will take the port number and a random `nonce` to pass along.  The RPC returns a success that the ping worked - but if the client didn't receive the expected `nonce` then there is another client on the network using that IP:Port combination - so libp2p can randomly select a port and then use NAT traversal on an unused port.  Ping is unlikely to be used for DDoS because it only sends a ping back to the IP address that sent the request - which should be enough to verify connectivity.

Now these two RPCs should resolve external routing conflicts because it allows a new node who is bootstrapping to see if there is already a node on the network who has claimed the default port number. If both nodes share the same IP address - they can still communicate as peers via the public interface even if they are on separate subnets. 

But there is another issue that you brought up which is shortcomings of mdns.  Two major reasons why mdns will fail is that the two clients are not on the same subnet - but still share the same NAT gateway, and the other being that mdns is filtered at the firewall level.  If mdns is filtered you can send a HELLO broadcast via another port, like libp2p's UDP 162 and see if any other nodes respond to the broadcast.  If they are on different subnets and share the same NAT gateway, they see the pattern above to establish connectivity via random port number.

**Comment by TheRook** [comment]
The real cool part of this solution is that every node - regardless of network topography will have the same unique connection string.  Every node in the network will have a consistent and unique public_ip:port tuple to connect, and at no point would there ever be a need to advertise a local IP address to any node.   This means that at most 65,535 libp2p nodes can all share the same public IP address - all of which can communicate with each-other and the global network.

**Comment by TheRook** [comment]
@Jorropo  please confirm that your issue is solved with the two new RPC calls.  Additionally, under no circumstance must libp2p node advertiser a RFC1918 address to over to a non-RFC1918 address as this is guaranteed to result in a routing error in all cases.

**Comment by TheRook** [comment]
@Jorropo   There there two additional methods on top of mDNS that can be used for any number of nodes to identify each-other on the same network.  Most notability, if two nodes who are not on the same subnet - but share the same NAT can communicate so long nodes are made aware of their external IP address. 

Let me be clear RFC1918 address cannot be routed to over the internet as per the RFC definition and must never be advertised under as this will always be an error condition.  This is always a bug, and there is a good patch here.

**Comment by TheRook** [comment]
@Jorropo please confirm.

**Comment by lidel** [comment]
I'm closing this, as too much changed in last 3-6 years for this issue to be useful 

The core problem -- private addresses leaking to the public DHT -- has been addressed across several releases:

  - **v0.5.0**: Dual WAN/LAN DHT ensures private addresses only go to LAN DHT.
  - **v0.28.0**: Kubo no longer keeps track of loopback addresses on LAN DHT or private addresses on WAN DHT. A `Routing.LoopbackAddressesOnLanDHT` config option exists for testing scenarios.
  - **Server profile** (`ipfs config profile apply server`): Explicitly filters all private/reserved ranges (including CGNAT `100.64.0.0/10`) from both `Swarm.AddrFilters` and there is also  `Addresses.NoAnnounce`. Recommended for machines with public IPv4 addresses that explicitly do not need LAN discovery.
  - **v0.40.0-rc1**: New `ipfs swarm addrs autonat` command allows users to self-diagnose address reachability via AutoNAT V2.  AutoNAT also makes sure any private undiallable addrs are not announced to other peers. 

  The original three-step plan from this META issue got implemented by other means (WAN/LAN separation + AutoNAT v2). Whether the `server` profile should be the default is a separate design questionm but IMO does not make a good default: filtering private addresses and disabling mDNS discovery by default would break LAN discovery for users running multiple IPFS nodes on the same network.

  If you still experience private address leakage in Kubo 0.40.0-rc1 or later (with or without the `server` profile), please open a new issue with specific reproduction steps. Same for any feature request -- should be filled as individual issues so we can triage them properly.
